Add lightbox tests covering theme.json configurations, including edge cases

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Since the UI for the lightbox has now been updated in https://github.com/WordPress/gutenberg/pull/57608 with extensive edge cases covered via https://github.com/WordPress/gutenberg/pull/59890#discussion_r1534695371, this PR reenables and updates the old lightbox tests while adding many new ones.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
We want to have test coverage on the lightbox to prevent regressions.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- It adds a test theme to sandbox all of the lightbox maneuvering, then toggles the various `theme.json` configurations via test plugins to reach various scenarios, testing both the UI and the frontend.
- It ensures lightbox settings in theme.json can be overridden via the global styles.
- For tests in the Site Editor, it adds blocks as needed then cleans them up.
- Since this PR is already big, it skips over various related tests, such as testing the lightbox markup and keyboard navigation, that I will handle in a separate PR.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
Run the tests locally with Docker running, using the command `npm run test:e2e:playwright -- image.spec.js --debug`

artemiomorales committed:
Rename lightbox trigger class to avoid conflicts with 3rd party code

artemiomorales committed:
Add label to Expand on Click button to facilitate testing

artemiomorales committed:
Add test themes

artemiomorales committed:
Add initial basic tests of theme.json lightbox settings

artemiomorales committed:
Add various test cases for theme.json configuration

artemiomorales committed:
Improve test descriptions and add additional checks to test cases

gziolo commented on packages/block-library/src/image/index.php:
```diff
@@ -204,7 +204,7 @@ function block_core_image_render_lightbox( $block_content, $block ) {
 		$img[0]
 		. '<button
-			class="lightbox-trigger"
+			class="wp-lightbox-trigger"
```
At this point, it's probably safer to leave the class name unchanged as sites might have styles in place that depend on the old class name.

artemiomorales replied:
Reverted 👍

gziolo commented on packages/block-editor/src/components/url-popover/image-url-input-ui.js:
```diff
@@ -369,6 +369,7 @@ const ImageURLInputUI = ( {
 											onSetLightbox( true );
 											stopEditLink();
 										} }
+										label={ __( 'Expand on click' ) }
```
What's the reasoning behind this change? The label is exactly the same as the text in the menu item.

artemiomorales replied:
I thought I would need this to access it by name, but realized while working on this PR that we don't need an explicit label to do that — this change has been reverted.

gziolo commented:
Nice job covering the scenarios with configuring the lightbox using the `theme.json` and the block settings.

> It adds four test themes containing different lightbox configurations, and toggles them on and off to reach desired edge cases.

It would be great to explore an alternative approach so don’t have to create 4 new test themes to only validate the settings. It doesn’t seem like a scalable approach, given how many settings are there in the `theme.json` and all the possible permutations. One alternative approach I was thinking about would be using WP filters. There are hooks that allow overriding settings in `theme.json` that might fit better there, but you would still need to activate them somehow, so that would probably still mean 4 plugins: https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/. A good side effect would be that those hooks would get some basic e2e test coverage.

gziolo commented on packages/e2e-test-utils-playwright/src/editor/update-post.ts:
```diff
@@ -0,0 +1,23 @@
+import type { Editor } from './index';
+
+/**
+ * Publishes the post, resolving once the request is complete (once a notice
```
```suggestion
 * Updates the post, resolving once the request is complete (once a notice
```

@Mamaduka or @kevin940726, can you confirm everything looks correct regarding adding a new method to the Editor class?

Mamaduka replied:
It does look good to me 👍 . The [general guidline](https://developer.wordpress.org/block-editor/contributors/code/testing-overview/e2e/#dont-overload-test-utils-inline-simple-utils) is to avoid adding new utilities when possible, but I think post updating is a common action.

artemiomorales replied:
Thanks for taking a look! The comment has been updated.

SantosGuillamot replied:
FYI: The same updatePost util was discussed in the tests for block bindings: [link](https://github.com/WordPress/gutenberg/pull/58676#discussion_r1478422153). In the end it was decided to keep it as a test util because it would require a deeper discussion.

Mamaduka replied:
Thanks for the link, @SantosGuillamot! I remembered a similar discussion but couldn't find it.

Now that more tests want to use similar action I think it's okay to introduce utility.

SantosGuillamot replied:
No problem! It makes sense to me as well 🙂 We might need to adapt it to the suggested changes there as the naming.

Mamaduka commented on packages/e2e-test-utils-playwright/src/editor/update-post.ts:
```diff
@@ -0,0 +1,23 @@
+ * @param this
+ */
+export async function updatePost( this: Editor ) {
+	const updateButton = this.page.getByRole( 'button', {
+		name: 'Update',
+	} );
+	await updateButton.click();
+
+	await this.page
+		.getByRole( 'button', {
+			name: 'Update',
+		} )
+		.waitFor();
```
I would recommend matching the helper from block binding e2e tests. Why?

* Uses scoped selectors.
* Waiting for notices turned out to be less flaky than waiting for label changes.

https://github.com/WordPress/gutenberg/blob/11f3096a6cb8f01afbd15ec4a440dade45ef1730/test/e2e/specs/editor/various/block-bindings.spec.js#L1933-L1942

kevin940726 replied:
It would probably be better to click (close) the dismiss notice too, so that we don't end up with duplicating ones. 👍

artemiomorales replied:
Done 👍

artemiomorales committed:
Add tests for overrides in global styles; add site editor compatibility for test themes

artemiomorales committed:
Add checks for lightbox UI in block toolbar in full site editor

artemiomorales committed:
Modify cleanup code so that it behaves more consistently

artemiomorales committed:
Revert "Rename lightbox trigger class to avoid conflicts with 3rd party code"

This reverts commit 59617d1cf2b9da09bc566c68d3254166a49f2612.

artemiomorales committed:
Revert "Add label to Expand on Click button to facilitate testing"

This reverts commit 0756190f81ca67a06c02a2850e3ffae8c4b4aea5.

artemiomorales committed:
Revise updatePost to behave more consistently

artemiomorales committed:
Update comment

artemiomorales committed:
Replace various test themes with plugins instead to reduce clutter

artemiomorales committed:
Fix PHPCS errors

artemiomorales committed:
Remove obsolete tests and skip others for now

artemiomorales commented:
> It would be great to explore an alternative approach so don’t have to create 4 new test themes to only validate the settings. It doesn’t seem like a scalable approach, given how many settings are there in the theme.json and all the possible permutations. One alternative approach I was thinking about would be using WP filters. There are hooks that allow overriding settings in theme.json that might fit better there, but you would still need to activate them somehow, so that would probably still mean 4 plugins: https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/. A good side effect would be that those hooks would get some basic e2e test coverage.

@gziolo Thanks for the suggestion! I removed the extra themes and moved the `theme.json` settings to plugins instead.

artemiomorales commented:
There is some flakiness in these tests that I'm still hoping to address, but I do think this is ready for reviewing.

SantosGuillamot commented:
I don't fully understand why we need a new theme for these tests. If we use plugins with filters, can't we use `emptytheme` for example?

SantosGuillamot commented:
> so that would probably still mean 4 plugins

In order to avoid having 4 plugins maybe it could have only 1 plugin that registers two site settings (with [register_setting](https://developer.wordpress.org/reference/functions/register_setting/)?): one for `allowEditing` and other for `enabled`. 

The plugin could use just one filter that can read from those settings and those settings can be updated in the tests by using the util `updateSiteSettings`: [example](https://github.com/WordPress/gutenberg/blob/d2191d4cf50d63c5f24885a8d3ab2c57a77e046c/test/e2e/specs/editor/blocks/post-comments-form.spec.js#L20-L22).

Could something like that work?

artemiomorales committed:
Remove erroneous code

artemiomorales commented:
> I don't fully understand why we need a new theme for these tests. If we use plugins with filters, can't we use emptytheme for example?

I initially included the new theme to avoid conflicting with other tests in case changes to the site templates somehow persisted — the cleanup code is a bit flaky when run locally. But you're right that we can do without this. Will continue looking at the cleanup code and remove the extra theme in a future push.

> In order to avoid having 4 plugins maybe it could have only 1 plugin that registers two site settings (with [register_setting](https://developer.wordpress.org/reference/functions/register_setting/)?): one for allowEditing and other for enabled.
>
> The plugin could use just one filter that can read from those settings and those settings can be updated in the tests by using the util updateSiteSettings: [example](https://github.com/WordPress/gutenberg/blob/d2191d4cf50d63c5f24885a8d3ab2c57a77e046c/test/e2e/specs/editor/blocks/post-comments-form.spec.js#L20-L22).
>
> Could something like that work?

Thanks for the suggestion! I'm actually thinking to simplify these tests to tackle just one `theme.json` configuration, so we wouldn't need all of these plugins anymore, at least to start. That being said, I'll look into this suggestion further as work on these tests continues.

artemiomorales commented:
This PR is very big and I hadn't been able to resolve the flakiness. It's been a few weeks now and looking at it with fresh eyes, I'd like to extract pieces of this PR to work on and review separately, which should be easier on me and reviewers, so will close this.
